### ISSUE TITLE
Performance issue when using instance profile credentials

### ISSUE THREAD

**Issue by gruntster** [aws/aws-sdk-net#718]
## Expected Behaviour
The retrieval of temporary security credentials from instance metadata should not cause performance degradation when using the SDK with multiple Tasks.

## Current Behaviour
The retrieval of temporary security credentials performs well when requests to AWS are made consecutively.  However, processing is significantly delayed when the refreshing of temporary security credentials is required whilst concurrent requests are performed.

## Steps to Reproduce (for bugs)
Run the following console application using an EC2 instance with an instance profile.  Amazon Linux, Ubuntu and Windows all exhibit the same behaviour.
```c#
using Amazon.SimpleNotificationService;
using Amazon.SimpleNotificationService.Model;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading.Tasks;

public class Program
{
    private static AmazonSimpleNotificationServiceClient client;

    public static void Main(string[] args)
    {
        client = new AmazonSimpleNotificationServiceClient();

        //PerformTest1().Wait();  // Uncomment this line to observe normal behaviour
        PerformTest2().Wait();
    }

    private static async Task PerformTest1()
    {
        var stopwatch = new Stopwatch();

        stopwatch.Start();

        await PublishMessage();

        stopwatch.Stop();

        Console.WriteLine($"Test 1 Elapsed time: {stopwatch.Elapsed.TotalSeconds} seconds");
    }

    private static async Task PerformTest2()
    {
        var stopwatch = new Stopwatch();
        var tasks = new List<Task<PublishResponse>>();

        stopwatch.Start();

        for (int counter = 0; counter < 20; counter++)
        {
            tasks.Add(Task.Run(async () => await PublishMessage()));
        }

        await Task.WhenAll(tasks);

        stopwatch.Stop();

        Console.WriteLine($"Test 2 Elapsed time: {stopwatch.Elapsed.TotalSeconds} seconds");
    }

    private static async Task<PublishResponse> PublishMessage() => await client.PublishAsync("arn:aws:sns:ap-southeast-2:111111111111:topic", "yo");
}
```
The `PerformTest1` method publishes a single message to an SNS topic.  The `PerformTest2` method publishes 20 messages to the SNS topic using Tasks.

Below is the execution time of the `PerformTest2` method when preceded by the `PerformTest1` method (to avoid rival fetching of security credentials):

> C:\ConsoleApp1>dotnet run
Test 1 Elapsed time: 0.2750664 seconds
Test 2 Elapsed time: **0.0763104 seconds**

Below is the execution time of the `PerformTest2` method without first running `PerformTest1`:

> C:\ConsoleApp1>dotnet run
Test 2 Elapsed time: **18.2262467 seconds**

## Context
We have various applications that do not make AWS calls for extended periods of time (e.g. an hour).  These applications are also required to handle sporadic load.  The retrieval of temporary credentials when performing concurrent processing can cause delays of over a minute compared to the normal processing time of around 200ms.

dotTrace has highlighted `Amazon.Runtime.InstanceProfileAWSCredentials.GenerateNewCredentials` as a major performance bottleneck for us.

## Your Environment
* AWSSDK.Core version used: `3.3.17.3`
* Service assembly and version used: `AWSSDK.SimpleNotificationService 3.3.0.21`
* Operating System and version: `Amazon Linux AMI 2017.03.1`, `Ubuntu Server 16.04 LTS`, `Microsoft Windows 2016 Datacenter edition`
* Visual Studio version: `Professional 2017 15.3.0 preview 7.0`
* Targeted .NET platform: `.NET Core`

## .NET Core Info
* .NET Core version used for development: `.NET Core 1.1.2 (x64)`
* .NET Core version installed in the environment where application runs: `.NET Core 1.1.2 (x64)`
* Output of `dotnet --info`: 
```
.NET Command Line Tools (1.0.4)

Product Information:
 Version:            1.0.4
 Commit SHA-1 hash:  af1e6684fd

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\1.0.4
```
* Contents of project.json/project.csproj: 
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp1.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="AWSSDK.SimpleNotificationService" Version="3.3.0.21" />
  </ItemGroup>

</Project>
```

**Comment by aliniacobv** [comment]
Any updates on this? 
We are encountering, possibly, the same issue, with the 1st publish request taking close to 10 seconds for a Lambda with a config of 128 MB. 

Code: 
`
public class ProvisioningMessage
    {
        public bool Ping { get; set; }
    }
public class Function
    {
        private const string SnsTopicArn = "messageReceivedArn";
        private string _snsTopicArn = Environment.GetEnvironmentVariable(SnsTopicArn);
        private readonly IAmazonSimpleNotificationService _snsService;
 
        public Function()
        {
            _snsService = new AmazonSimpleNotificationServiceClient();
        }
 
        public Function(IAmazonSimpleNotificationService amazonSimpleNotificationService)
        {
            _snsService = amazonSimpleNotificationService;
        }
 
        public async Task FunctionHandler(ProvisioningMessage provisioningMessage, ILambdaContext context)
        {
            if (provisioningMessage.Ping)
            {
                return;
            }
 
            Stopwatch sw = new Stopwatch();
            sw.Start();
            var snsResult = await _snsService.PublishAsync(
                    new PublishRequest
                    {
                        TopicArn = _snsTopicArn,
                        Message = JsonConvert.SerializeObject(provisioningMessage, new JsonSerializerSettings() { NullValueHandling = NullValueHandling.Ignore })
                    }
                );
            context.Logger.LogLine("Elapsed: " + sw.ElapsedMilliseconds);
 
            if (snsResult.HttpStatusCode != System.Net.HttpStatusCode.OK)
                throw new Exception($"PublishAsync returned HTTP code: {snsResult.HttpStatusCode}.");
        }
    }`

Sampling with CloudWatch logs:
1. First call with payload (cold start): { "Ping":true}. Logs:
START RequestId: e41d73e7-d991-11e7-9a99-8302147c44ca Version: $LATEST
END RequestId: e41d73e7-d991-11e7-9a99-8302147c44ca
REPORT RequestId: e41d73e7-d991-11e7-9a99-8302147c44ca	Duration: 4329.36 ms	Billed Duration: 4400 ms 	Memory Size: 128 MB	Max Memory Used: 37 MB

2. Second call (a few seconds later), with payload: { "Ping":false} 
START RequestId: 0e4365a2-d992-11e7-9aab-a52fb60edb50 Version: $LATEST
Elapsed: 8442
END RequestId: 0e4365a2-d992-11e7-9aab-a52fb60edb50
REPORT RequestId: 0e4365a2-d992-11e7-9aab-a52fb60edb50	Duration: 8490.09 ms	Billed Duration: 8500 ms 	Memory Size: 128 MB	Max Memory Used: 47 MB	

3. Third call (a few seconds later), with payload: { "Ping":false} 
START RequestId: 2e3f6ad4-d992-11e7-a8b5-4b7caee919ff Version: $LATEST
Elapsed: 383
END RequestId: 2e3f6ad4-d992-11e7-a8b5-4b7caee919ff
REPORT RequestId: 2e3f6ad4-d992-11e7-a8b5-4b7caee919ff	Duration: 384.71 ms	Billed Duration: 400 ms 	Memory Size: 128 MB	Max Memory Used: 47 MB	

4. Forth call (a few seconds later), with payload: { "Ping":false} 
START RequestId: 424a5632-d992-11e7-9fa2-d352ca9e984d Version: $LATEST
Elapsed: 189
END RequestId: 424a5632-d992-11e7-9fa2-d352ca9e984d
REPORT RequestId: 424a5632-d992-11e7-9fa2-d352ca9e984d	Duration: 190.43 ms	Billed Duration: 200 ms 	Memory Size: 128 MB	Max Memory Used: 47 MB

**Comment by sstevenkang** [comment]
Have you tried updating AWSSDK.Core to the latest?  We have addressed this in [3.3.330.2](https://github.com/aws/aws-sdk-net/blob/master/SDK.CHANGELOG.md#333302-2018-08-03-2218-utc)

**Comment by klaytaybai** [comment]
Closing due to inactivity and a tested solution
